{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d69c6e75-65e8-4698-9f8b-b4ffd7ec0a67",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_42845/1937594232.py:4: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import pkg_resources\n",
    "from geovar import *\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ed14ade7-627c-44ed-a632-94bfc420fec9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#data_path = pkg_resources.resource_filename(\"geovar\", \"data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "336b9877-9ada-42fe-9886-e23da33005b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/share/hennlab/progs/miniconda3/envs/geovar/lib/python3.13/site-packages/geovar/data/integrated_call_samples_v3.20130502.1kg_superpops.panel'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # Filepath to the VCF File\n",
    "# vcf_file = \"{}/new_1kg_nygc.chr22.biallelic_snps.filt.n5000.vcf.gz\".format(data_path)\n",
    "\n",
    "# # Filepath to the population panel file\n",
    "# population_panel = \"{}/integrated_call_samples_v3.20130502.1kg_superpops.panel\".format(data_path)\n",
    "\n",
    "# population_panel\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3340cf69-b2ba-447b-a389-9571b77d3fe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the base data path\n",
    "data_path = \"/share/hennlab\"\n",
    "\n",
    "# Filepath to the VCF file\n",
    "vcf_file = \"{}/data/genomes/CAAPA_freeze2_PHASED_common_rare/snps/chr21.20240612_Freeze2.common_rare_phased.snps.vcf.gz\".format(data_path)\n",
    "\n",
    "# Filepath to the population panel file\n",
    "population_panel = \"{}/projects/CAAPA2_functional_annotation/geovar/data/cleaned_population_panel.txt\".format(data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d0394e86-5643-446d-9d60-539b056e8187",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/share/hennlab/data/genomes/CAAPA_freeze2_PHASED_common_rare/snps/chr21.20240612_Freeze2.common_rare_phased.snps.vcf.gz'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "population_panel\n",
    "vcf_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "113bdc30-7ce9-4e5d-bacf-0e419bce1204",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>pop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFA0001</td>\n",
       "      <td>African-American</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AFA0002</td>\n",
       "      <td>African-American</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AFA0003</td>\n",
       "      <td>African-American</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AFA0004</td>\n",
       "      <td>African-American</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFA0005</td>\n",
       "      <td>African-American</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3449</th>\n",
       "      <td>YOR0046</td>\n",
       "      <td>West_Central_African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3450</th>\n",
       "      <td>YOR0047</td>\n",
       "      <td>West_Central_African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3451</th>\n",
       "      <td>YOR0048</td>\n",
       "      <td>West_Central_African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3452</th>\n",
       "      <td>YOR0049</td>\n",
       "      <td>West_Central_African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3453</th>\n",
       "      <td>YOR0050</td>\n",
       "      <td>West_Central_African</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3454 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample                   pop\n",
       "0     AFA0001      African-American\n",
       "1     AFA0002      African-American\n",
       "2     AFA0003      African-American\n",
       "3     AFA0004      African-American\n",
       "4     AFA0005      African-American\n",
       "...       ...                   ...\n",
       "3449  YOR0046  West_Central_African\n",
       "3450  YOR0047  West_Central_African\n",
       "3451  YOR0048  West_Central_African\n",
       "3452  YOR0049  West_Central_African\n",
       "3453  YOR0050  West_Central_African\n",
       "\n",
       "[3454 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reading the population dataframe\n",
    "pop_df = read_pop_panel(population_panel)\n",
    "pop_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a77e328-0d45-4867-bc7c-2defe6b52362",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1665246it [37:43, 748.02it/s]"
     ]
    }
   ],
   "source": [
    "# Writing out VCF to a Frequency Table\n",
    "af_df = vcf_to_freq_table(vcf_file, pop_df=pop_df, outfile=\"/share/hennlab/projects/CAAPA2_functional_annotation/geovar/data/freq_mat_file.csv\", minor_allele=True)\n",
    "\n",
    "# Print the beginning of the allele frequency table\n",
    "af_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1f9c3c95-5df1-476a-ae7e-24fd6c51227d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_42845/777710661.py:7: DtypeWarning: Columns (13,18,19) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  neutral_snp_df = pd.read_csv(neutral_snp_file, sep=\"\\t\")   # Assuming TSV is tab-separated\n",
      "/tmp/ipykernel_42845/777710661.py:8: DtypeWarning: Columns (19) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  conserved_snp_df = pd.read_csv(conserved_snp_file, sep=\"\\t\")   # Assuming TSV is tab-separated\n"
     ]
    }
   ],
   "source": [
    " \n",
    "# here im creating two subset freqmat files\n",
    "# File paths\n",
    "freq_file = \"/share/hennlab/projects/CAAPA2_functional_annotation/geovar/data/freq_mat_file.csv\"\n",
    "neutral_snp_file = \"/share/hennlab/projects/CAAPA2_functional_annotation/dbNSFP_v4.9/scripts/neutral_chr_21_snps.tsv\"\n",
    "conserved_snp_file = \"/share/hennlab/projects/CAAPA2_functional_annotation/dbNSFP_v4.9/scripts/conserved_chr_21_snps.tsv\"\n",
    "# Load files\n",
    "freq_df = pd.read_csv(freq_file, sep=\"\\t\")  # Assuming CSV is comma-separated\n",
    "neutral_snp_df = pd.read_csv(neutral_snp_file, sep=\"\\t\")   # Assuming TSV is tab-separated\n",
    "conserved_snp_df = pd.read_csv(conserved_snp_file, sep=\"\\t\")   # Assuming TSV is tab-separated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d441a2e2-ffcf-424d-82cf-25bf6c1b8eca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method NDFrame.head of            CHR       SNP A1 A2  Chromosome  Position   GerpN   GerpS  \\\n",
       "0        chr21   5239594  C  T          21   5239594  3.7200  1.9000   \n",
       "1        chr21   5239599  T  C          21   5239599  3.7200  1.1400   \n",
       "2        chr21   5239603  T  A          21   5239603  3.5600 -0.2490   \n",
       "3        chr21   5239607  C  G          21   5239607  3.5600 -1.9900   \n",
       "4        chr21   5239608  T  C          21   5239608  3.6800 -0.1860   \n",
       "...        ...       ... .. ..         ...       ...     ...     ...   \n",
       "1514593  chr21  46679245  T  C          21  46679245  0.0615  0.0615   \n",
       "1514594  chr21  46679304  A  C          21  46679304  0.0615  0.0615   \n",
       "1514595  chr21  46679324  A  G          21  46679324  0.1100  0.1100   \n",
       "1514596  chr21  46679365  C  G          21  46679365  0.1100  0.1100   \n",
       "1514597  chr21  46679372  C  T          21  46679372  0.1100  0.1100   \n",
       "\n",
       "         Priphylop  Mamphylop  ...  ApcProteinFunctionV3  FunseqDescription  \\\n",
       "0           -0.567      0.241  ...              2.969487                NaN   \n",
       "1            0.436      0.301  ...              2.969487                NaN   \n",
       "2           -0.408      0.032  ...              2.969487                NaN   \n",
       "3            0.389     -0.674  ...              2.969487                NaN   \n",
       "4            0.389     -0.041  ...              2.969487                NaN   \n",
       "...            ...        ...  ...                   ...                ...   \n",
       "1514593      0.112      0.112  ...              2.969487          noncoding   \n",
       "1514594      0.114      0.113  ...              2.969487          noncoding   \n",
       "1514595      0.174      0.173  ...              2.969487          noncoding   \n",
       "1514596      0.155      0.156  ...              2.969487          noncoding   \n",
       "1514597      0.155      0.156  ...              2.969487          noncoding   \n",
       "\n",
       "         GenecodeComprehensiveCategory          GenecodeComprehensiveInfo  \\\n",
       "0                       ncRNA_intronic                         AC079801.1   \n",
       "1                       ncRNA_intronic                         AC079801.1   \n",
       "2                       ncRNA_intronic                         AC079801.1   \n",
       "3                       ncRNA_intronic                         AC079801.1   \n",
       "4                       ncRNA_intronic                         AC079801.1   \n",
       "...                                ...                                ...   \n",
       "1514593                     intergenic  PRMT2(dist=14121),NONE(dist=NONE)   \n",
       "1514594                     intergenic  PRMT2(dist=14180),NONE(dist=NONE)   \n",
       "1514595                     intergenic  PRMT2(dist=14200),NONE(dist=NONE)   \n",
       "1514596                     intergenic  PRMT2(dist=14241),NONE(dist=NONE)   \n",
       "1514597                     intergenic  PRMT2(dist=14248),NONE(dist=NONE)   \n",
       "\n",
       "            AfAfr     AfAmr  Clndn  Clndnincl Gerpscale1 Gerpscale2  \n",
       "0             NaN       NaN    NaN        NaN    neutral    neutral  \n",
       "1        0.000000  0.000146    NaN        NaN    neutral    neutral  \n",
       "2        0.000214  0.000000    NaN        NaN    neutral    neutral  \n",
       "3             NaN       NaN    NaN        NaN    neutral    neutral  \n",
       "4        0.000000  0.000000    NaN        NaN    neutral    neutral  \n",
       "...           ...       ...    ...        ...        ...        ...  \n",
       "1514593  0.000071  0.000000    NaN        NaN    neutral    neutral  \n",
       "1514594       NaN       NaN    NaN        NaN    neutral    neutral  \n",
       "1514595  0.000333  0.000659    NaN        NaN    neutral    neutral  \n",
       "1514596  0.000857  0.000000    NaN        NaN    neutral    neutral  \n",
       "1514597       NaN       NaN    NaN        NaN    neutral    neutral  \n",
       "\n",
       "[1514598 rows x 22 columns]>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#freq_df.head\n",
    "neutral_snp_df.head\n",
    "#conserved_snp_df.head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dade6ec-3987-427a-90f5-2e604d0a0678",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter based on SNP column (assuming SNP column is named 'SNP' in both files)\n",
    "filtered_df_neutral = freq_df[freq_df[\"SNP\"].isin(neutral_snp_df[\"SNP\"])]\n",
    "filtered_df_conserved = freq_df[freq_df[\"SNP\"].isin(conserved_snp_df[\"SNP\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ca17fd9f-91a4-4d87-803b-279f091a193e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1514521\n",
      "152200\n"
     ]
    }
   ],
   "source": [
    "print(len(filtered_df_neutral))    \n",
    "print(len(filtered_df_conserved))  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f1db6ad4-532a-4927-a81d-45972741cfaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the filtered DataFrame\n",
    "filtered_df_neutral.to_csv(\"/share/hennlab/projects/CAAPA2_functional_annotation/geovar/data/freq_mat_chr21_neutral_filtered.csv\", index=False)\n",
    "filtered_df_conserved.to_csv(\"/share/hennlab/projects/CAAPA2_functional_annotation/geovar/data/freq_mat_chr21_conserved_filtered.csv\", index=False)\n",
    "\n",
    "#print(f\"Filtered file saved with {len(filtered_df)} SNPs.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9d7b9360-6ce7-4828-9b90-9b919ddf96cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1667683"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(af_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "16cc3990-58c6-4b57-b239-65f2e4890200",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating the GeoVar Object\n",
    "geovar_test = GeoVar()\n",
    "geovar_neutral = GeoVar()\n",
    "geovar_conserved = GeoVar()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "36260ebf-342a-45fb-8df4-09fd15488e81",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding in the frequency file (all of it)\n",
    "geovar_test.add_freq_mat(freq_mat_file=\"/share/hennlab/projects/CAAPA2_functional_annotation/geovar/data/freq_mat_file.csv\")\n",
    "geovar_neutral.add_freq_mat(freq_mat_file=\"/share/hennlab/projects/CAAPA2_functional_annotation/geovar/data/freq_mat_chr21_neutral_filtered.csv\")\n",
    "geovar_conserved.add_freq_mat(freq_mat_file=\"/share/hennlab/projects/CAAPA2_functional_annotation/geovar/data/freq_mat_chr21_conserved_filtered.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "2459c076-3b20-40ef-9f1b-f8a7d723e4e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GeoVar\n",
      "number of variants: 1667683\n",
      "number of pops: 13\n",
      "pops: African-American,Afro-Caribbean,Afro-South_American,Asian,Back-to-African,Eastern_Bantu,European,Fulani,Indigenous_American,Khoe-San,Northeast_African,Rainforest_Hunter-Gatherer,West_Central_African\n",
      "allele freq bins: (0, 0),(0, 0.05),(0.05, 1.0)\n"
     ]
    }
   ],
   "source": [
    "# Generate a geovar binning with the binning we used in our paper\n",
    "geovar_test.geovar_binning()\n",
    "\n",
    "# Printing details about the GeoVar object\n",
    "print(geovar_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "3e80f43a-abd7-492b-a524-a99bef29d6cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "import gzip\n",
    "#load in James Kitchens Function\n",
    "\n",
    "def geovar_codes_streaming_fixed(geovar_obj, freq_mat_file):\n",
    "    \"\"\"Version of GeoVar code generation algorithm that streams through file to avoid memory overflow.\n",
    "    Args:\n",
    "        freq_mat_file (:obj:`string`): filepath to\n",
    "        frequency table file (see example notebook for formatting).\n",
    "    \"\"\"\n",
    "    assert geovar_obj.bins is not None\n",
    "    geovar_codes = []\n",
    "    # Setting up the testing bins\n",
    "    test_bins = np.array([x[1] for x in geovar_obj.bins])\n",
    "    with gzip.open(freq_mat_file,'r') as f:\n",
    "        header = f.readline()\n",
    "        # Take the population labels currently\n",
    "        pops = np.array(header.split()[6:])\n",
    "        geovar_obj.pops = pops\n",
    "        for line in tqdm(f):\n",
    "            # Split after the 6th column ...\n",
    "            maf_vector = np.array(line.split()[6:]).astype(np.float64)\n",
    "            cur_geovar = np.digitize(maf_vector, test_bins, right=True)\n",
    "            cur_geovar_code = \"\".join([str(i) for i in cur_geovar])\n",
    "            geovar_codes.append(cur_geovar_code)\n",
    "    # Setting the variables here\n",
    "    geovar_obj.geovar_codes = np.array(geovar_codes)\n",
    "    geovar_obj.n_variants = geovar_obj.geovar_codes.size\n",
    "    geovar_obj.n_populations = geovar_obj.pops.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "d90c565e-f5bf-40a9-aa51-fcb7bf31fa12",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1667683it [01:04, 25737.07it/s]\n"
     ]
    }
   ],
   "source": [
    "geovar = GeoVar(bins=[(0,0), (0,0.01), (0.01,0.05), (0.05,0.1), (0.1,1.0)])\n",
    "geovar_codes_streaming_fixed(geovar_obj=geovar, freq_mat_file=\"/share/hennlab/projects/CAAPA2_functional_annotation/geovar/data/freq_mat_file.csv.gz\")\n",
    "counts_output = geovar.count_geovar_codes()\n",
    "geovar_code_counts = pd.DataFrame({\"codes\":counts_output[0],\"counts\":counts_output[1]})\n",
    "geovar_code_counts.to_csv(\"/share/hennlab/projects/CAAPA2_functional_annotation/geovar/data/caapa_chr21.txt\", sep=' ', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "dd5b6f04-692d-406b-a2d7-c1765925d90c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                codes  counts\n",
      "0       0000000000001  189451\n",
      "1       0000000000010   14886\n",
      "2       0000000000011    4740\n",
      "3       0000000000020    7949\n",
      "4       0000000000021    2535\n",
      "...               ...     ...\n",
      "106060  4444444444433     196\n",
      "106061  4444444444434    1610\n",
      "106062  4444444444442       3\n",
      "106063  4444444444443     206\n",
      "106064  4444444444444   33803\n",
      "\n",
      "[106065 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "print(geovar_code_counts)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:geovar]",
   "language": "python",
   "name": "conda-env-geovar-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
