{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "facc4fef-6bff-483e-a5db-81630dc21872",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2642973/4215322703.py:4: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources #pip install setuptools\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import pkg_resources #pip install setuptools\n",
    "from geovar import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9b1b3095-b5d2-40c0-aa2d-96985ad0a02c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>pop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFA0001</td>\n",
       "      <td>African-American</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AFA0002</td>\n",
       "      <td>African-American</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AFA0003</td>\n",
       "      <td>African-American</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AFA0004</td>\n",
       "      <td>African-American</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFA0005</td>\n",
       "      <td>African-American</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3449</th>\n",
       "      <td>YOR0046</td>\n",
       "      <td>West_Central_African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3450</th>\n",
       "      <td>YOR0047</td>\n",
       "      <td>West_Central_African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3451</th>\n",
       "      <td>YOR0048</td>\n",
       "      <td>West_Central_African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3452</th>\n",
       "      <td>YOR0049</td>\n",
       "      <td>West_Central_African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3453</th>\n",
       "      <td>YOR0050</td>\n",
       "      <td>West_Central_African</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3454 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample                   pop\n",
       "0     AFA0001      African-American\n",
       "1     AFA0002      African-American\n",
       "2     AFA0003      African-American\n",
       "3     AFA0004      African-American\n",
       "4     AFA0005      African-American\n",
       "...       ...                   ...\n",
       "3449  YOR0046  West_Central_African\n",
       "3450  YOR0047  West_Central_African\n",
       "3451  YOR0048  West_Central_African\n",
       "3452  YOR0049  West_Central_African\n",
       "3453  YOR0050  West_Central_African\n",
       "\n",
       "[3454 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define the base data path\n",
    "data_path = \"/quobyte/bmhenngrp/from-lssc0/\"\n",
    "\n",
    "# Filepath to the VCF file (THIS IS CURRENTLY CHR 21 but it will be expannded to ALL chromosomes when caapap v2 is ready)\n",
    "#vcf_file = \"{}/data/genomes/CAAPA_freeze2_PHASED_common_rare/snps/chr21.20240612_Freeze2.common_rare_phased.snps.vcf.gz\".format(data_path)\n",
    "vcf_file = \"{}data/genomes/2025_CAAPA_Freeze2B/combined/Freeze2.BiallelicSNPs_shapeit5_Phased_common_rare_allchr.vcf.gz\".format(data_path)\n",
    "#vcf_file = \"{}data/genomes/2025_CAAPA_Freeze2B/combined/Freeze2.BiallelicSNPs_shapeit5_Phased_common_rare_chr22.vcf.gz\".format(data_path)\n",
    "\n",
    "# Filepath to the population panel file ( 2 COLUMN tab delimited file with sample id and pop name)\n",
    "population_panel = \"{}/projects/CAAPA2_functional_annotation/geovar/data/cleaned_population_panel.txt\".format(data_path)\n",
    "\n",
    "# Reading the population dataframe\n",
    "pop_df = read_pop_panel(population_panel)\n",
    "pop_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dd16e22-c647-4f91-9fa6-031193bdb823",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Writing out VCF to a Frequency Table (*Majestiq will use the ouptut from this as side project)\n",
    "af_df = vcf_to_freq_table(vcf_file, pop_df=pop_df, outfile=\"/quobyte/bmhenngrp/from-lssc0/projects/CAAPA2_functional_annotation/geovar/results/freq_mat_file_v2.csv\", minor_allele=True)\n",
    "\n",
    "# Print the beginning of the allele frequency table\n",
    "af_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bd0cd870-928d-419a-b3eb-1d75c93d0b28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# File paths\n",
    "freq_file = \"/quobyte/bmhenngrp/from-lssc0/projects/CAAPA2_functional_annotation/geovar/results/freq_mat_file_v2.csv\"\n",
    "# Load files\n",
    "freq_df = pd.read_csv(freq_file, sep=\"\\t\")  # Assuming CSV is comma-separated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ed9b2e8d-06e5-41fc-90e9-e5e83aa87a66",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "18ceb69e-0aa4-4942-9550-6b0dad2ffa44",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Creating the GeoVar Object\n",
    "geovar_test = GeoVar()\n",
    "\n",
    "\n",
    "# Adding in the frequency file (all of it)\n",
    "#geovar_test.add_freq_mat(freq_mat_file=\"/quobyte/bmhenngrp/from-lssc0/projects/CAAPA2_functional_annotation/geovar/results/freq_mat_file_v2.csv\")\n",
    "geovar_test.add_freq_mat(freq_mat_file=\"/quobyte/bmhenngrp/from-lssc0/projects/CAAPA2_functional_annotation/geovar/results/freq_mat_file_allchr.csv\")\n",
    "\n",
    "\n",
    "# input_file = \"/quobyte/bmhenngrp/from-lssc0/projects/CAAPA2_functional_annotation/geovar/results/freq_mat_file_v2.csv\"\n",
    "# output_file = input_file + \".gz\"\n",
    "\n",
    "# with open(input_file, \"rb\") as f_in:\n",
    "#     with gzip.open(output_file, \"wb\") as f_out:\n",
    "#         shutil.copyfileobj(f_in, f_out)\n",
    "\n",
    "# print(\"Gzipped:\", output_file)\n",
    "#once this is done create a zipped version of the freq_mat_file in the results folder to use downstream"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "80e6b067-798d-493f-9aa4-5db7d30d70b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GeoVar\n",
      "number of variants: 39088824\n",
      "number of pops: 13\n",
      "pops: African-American,Afro-Caribbean,Afro-South_American,Asian,Back-to-African,Eastern_Bantu,European,Fulani,Indigenous_American,Khoe-San,Northeast_African,Rainforest_Hunter-Gatherer,West_Central_African\n",
      "allele freq bins: (0, 0),(0, 0.05),(0.05, 1.0)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "39088824it [09:30, 68527.57it/s]\n"
     ]
    }
   ],
   "source": [
    "# Generate a geovar binning with the binning we used in our paper\n",
    "geovar_test.geovar_binning()\n",
    "\n",
    "# Printing details about the GeoVar object\n",
    "print(geovar_test)\n",
    "\n",
    "from tqdm import tqdm\n",
    "import gzip\n",
    "\n",
    "#load in James Kitchens Function from github \n",
    "def geovar_codes_streaming_fixed(geovar_obj, freq_mat_file):\n",
    "    \"\"\"Version of GeoVar code generation algorithm that streams through file to avoid memory overflow.\n",
    "    Args:\n",
    "        freq_mat_file (:obj:`string`): filepath to\n",
    "        frequency table file (see example notebook for formatting).\n",
    "    \"\"\"\n",
    "    assert geovar_obj.bins is not None\n",
    "    geovar_codes = []\n",
    "    # Setting up the testing bins\n",
    "    test_bins = np.array([x[1] for x in geovar_obj.bins])\n",
    "    with gzip.open(freq_mat_file,'r') as f:\n",
    "        header = f.readline()\n",
    "        # Take the population labels currently\n",
    "        pops = np.array(header.split()[6:])\n",
    "        geovar_obj.pops = pops\n",
    "        for line in tqdm(f):\n",
    "            # Split after the 6th column ...\n",
    "            maf_vector = np.array(line.split()[6:]).astype(np.float64)\n",
    "            cur_geovar = np.digitize(maf_vector, test_bins, right=True)\n",
    "            cur_geovar_code = \"\".join([str(i) for i in cur_geovar])\n",
    "            geovar_codes.append(cur_geovar_code)\n",
    "    # Setting the variables here\n",
    "    geovar_obj.geovar_codes = np.array(geovar_codes)\n",
    "    geovar_obj.n_variants = geovar_obj.geovar_codes.size\n",
    "    geovar_obj.n_populations = geovar_obj.pops.size\n",
    "\n",
    "geovar = GeoVar(bins=[(0,0), (0,0.01), (0.01,0.05), (0.05,0.1), (0.1,1.0)])\n",
    "geovar_codes_streaming_fixed(geovar_obj=geovar, freq_mat_file=\"/quobyte/bmhenngrp/from-lssc0/projects/CAAPA2_functional_annotation/geovar/results/freq_mat_file_allchr.csv.gz\") \n",
    "counts_output = geovar.count_geovar_codes()\n",
    "geovar_code_counts = pd.DataFrame({\"codes\":counts_output[0],\"counts\":counts_output[1]})\n",
    "geovar_code_counts.to_csv(\"/quobyte/bmhenngrp/from-lssc0/projects/CAAPA2_functional_annotation/geovar/results/allchr.txt\", sep=' ', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5bbd2f2-35b2-455f-b9a4-ec55ef2f5b58",
   "metadata": {},
   "outputs": [],
   "source": [
    " \n",
    " "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:python_geovar]",
   "language": "python",
   "name": "conda-env-python_geovar-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
