{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6442c5e6-70eb-41ac-b622-1c6e2e34a707",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import pkg_resources #pip install setuptools\n",
    "from geovar import *\n",
    "\n",
    "\n",
    "# Define the base data path\n",
    "data_path = \"/quobyte/bmhenngrp/from-lssc0/\"\n",
    "\n",
    "# Filepath to the VCF file (THIS IS CURRENTLY CHR 21 but it will be expannded to ALL chromosomes when caapap v2 is ready)\n",
    "#vcf_file = \"{}/data/genomes/CAAPA_freeze2_PHASED_common_rare/snps/chr21.20240612_Freeze2.common_rare_phased.snps.vcf.gz\".format(data_path)\n",
    "vcf_file = \"{}data/genomes/2025_CAAPA_Freeze2B/combined/Freeze2.BiallelicSNPs_shapeit5_Phased_common_rare_allchr.vcf.gz\".format(data_path)\n",
    "# Filepath to the population panel file ( 2 COLUMN tab delimited file with sample id and pop name)\n",
    "population_panel = \"{}/projects/CAAPA2_functional_annotation/geovar/data/cleaned_population_panel.txt\".format(data_path)\n",
    "\n",
    "# Reading the population dataframe\n",
    "pop_df = read_pop_panel(population_panel)\n",
    "pop_df\n",
    "\n",
    "# Writing out VCF to a Frequency Table (*Majestiq will use the ouptut from this as side project)\n",
    "af_df = vcf_to_freq_table(vcf_file, pop_df=pop_df, outfile=\"/quobyte/bmhenngrp/from-lssc0/projects/CAAPA2_functional_annotation/geovar/results/freq_mat_file_allchr.csv\", minor_allele=True)\n",
    "\n",
    "# Print the beginning of the allele frequency table\n",
    "af_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bc92af5-0a3d-47a7-a894-b5d94f1b526b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2126455/2567765615.py:4: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources #pip install setuptools\n",
      "/quobyte/bmhenngrp/conda_envs/python_geovar/lib/python3.14/site-packages/geovar/utils.py:52: UserWarning: Sample SHK4014new does not have a population label!\n",
      "  warnings.warn(f\"Sample {s} does not have a population label!\", UserWarning)\n",
      "/quobyte/bmhenngrp/conda_envs/python_geovar/lib/python3.14/site-packages/geovar/utils.py:52: UserWarning: Sample FUL6059new does not have a population label!\n",
      "  warnings.warn(f\"Sample {s} does not have a population label!\", UserWarning)\n",
      "/quobyte/bmhenngrp/conda_envs/python_geovar/lib/python3.14/site-packages/geovar/utils.py:52: UserWarning: Sample BRB6108new does not have a population label!\n",
      "  warnings.warn(f\"Sample {s} does not have a population label!\", UserWarning)\n",
      "/quobyte/bmhenngrp/conda_envs/python_geovar/lib/python3.14/site-packages/geovar/utils.py:52: UserWarning: Sample HG01601 does not have a population label!\n",
      "  warnings.warn(f\"Sample {s} does not have a population label!\", UserWarning)\n",
      "/quobyte/bmhenngrp/conda_envs/python_geovar/lib/python3.14/site-packages/geovar/utils.py:52: UserWarning: Sample KIG0064 does not have a population label!\n",
      "  warnings.warn(f\"Sample {s} does not have a population label!\", UserWarning)\n",
      "/quobyte/bmhenngrp/conda_envs/python_geovar/lib/python3.14/site-packages/geovar/utils.py:52: UserWarning: Sample KIG0006 does not have a population label!\n",
      "  warnings.warn(f\"Sample {s} does not have a population label!\", UserWarning)\n",
      "/quobyte/bmhenngrp/conda_envs/python_geovar/lib/python3.14/site-packages/geovar/utils.py:52: UserWarning: Sample KIG0030 does not have a population label!\n",
      "  warnings.warn(f\"Sample {s} does not have a population label!\", UserWarning)\n",
      "/quobyte/bmhenngrp/conda_envs/python_geovar/lib/python3.14/site-packages/geovar/utils.py:52: UserWarning: Sample KIG0067 does not have a population label!\n",
      "  warnings.warn(f\"Sample {s} does not have a population label!\", UserWarning)\n",
      "/quobyte/bmhenngrp/conda_envs/python_geovar/lib/python3.14/site-packages/geovar/utils.py:52: UserWarning: Sample KIG0014 does not have a population label!\n",
      "  warnings.warn(f\"Sample {s} does not have a population label!\", UserWarning)\n",
      "/quobyte/bmhenngrp/conda_envs/python_geovar/lib/python3.14/site-packages/geovar/utils.py:52: UserWarning: Sample KIG0009 does not have a population label!\n",
      "  warnings.warn(f\"Sample {s} does not have a population label!\", UserWarning)\n",
      "/quobyte/bmhenngrp/conda_envs/python_geovar/lib/python3.14/site-packages/geovar/utils.py:52: UserWarning: Sample KIG0010 does not have a population label!\n",
      "  warnings.warn(f\"Sample {s} does not have a population label!\", UserWarning)\n",
      "/quobyte/bmhenngrp/conda_envs/python_geovar/lib/python3.14/site-packages/geovar/utils.py:52: UserWarning: Sample NA19145 does not have a population label!\n",
      "  warnings.warn(f\"Sample {s} does not have a population label!\", UserWarning)\n",
      "1586189it [13:50, 1911.13it/s]"
     ]
    }
   ],
   "source": [
    "# File paths\n",
    "freq_file = \"/quobyte/bmhenngrp/from-lssc0/projects/CAAPA2_functional_annotation/geovar/data/freq_mat_file_v2.csv\"\n",
    " \n",
    "# Load files\n",
    "freq_df = pd.read_csv(freq_file, sep=\"\\t\")  # Assuming CSV is comma-separated\n",
    "\n",
    "# Creating the GeoVar Object\n",
    "geovar_test = GeoVar()\n",
    "\n",
    "\n",
    "# Adding in the frequency file (all of it)\n",
    "geovar_test.add_freq_mat(freq_mat_file=\"/quobyte/bmhenngrp/from-lssc0/projects/CAAPA2_functional_annotation/geovar/results/freq_mat_file_v2.csv\")\n",
    "\n",
    "#once this is done create a zipped version of the freq_mat_file in the results folder to use downstream\n",
    "\n",
    "\n",
    "# Generate a geovar binning with the binning we used in our paper\n",
    "geovar_test.geovar_binning()\n",
    "\n",
    "# Printing details about the GeoVar object\n",
    "print(geovar_test)\n",
    "\n",
    "from tqdm import tqdm\n",
    "import gzip\n",
    "\n",
    "#load in James Kitchens Function from github \n",
    "def geovar_codes_streaming_fixed(geovar_obj, freq_mat_file):\n",
    "    \"\"\"Version of GeoVar code generation algorithm that streams through file to avoid memory overflow.\n",
    "    Args:\n",
    "        freq_mat_file (:obj:`string`): filepath to\n",
    "        frequency table file (see example notebook for formatting).\n",
    "    \"\"\"\n",
    "    assert geovar_obj.bins is not None\n",
    "    geovar_codes = []\n",
    "    # Setting up the testing bins\n",
    "    test_bins = np.array([x[1] for x in geovar_obj.bins])\n",
    "    with gzip.open(freq_mat_file,'r') as f:\n",
    "        header = f.readline()\n",
    "        # Take the population labels currently\n",
    "        pops = np.array(header.split()[6:])\n",
    "        geovar_obj.pops = pops\n",
    "        for line in tqdm(f):\n",
    "            # Split after the 6th column ...\n",
    "            maf_vector = np.array(line.split()[6:]).astype(np.float64)\n",
    "            cur_geovar = np.digitize(maf_vector, test_bins, right=True)\n",
    "            cur_geovar_code = \"\".join([str(i) for i in cur_geovar])\n",
    "            geovar_codes.append(cur_geovar_code)\n",
    "    # Setting the variables here\n",
    "    geovar_obj.geovar_codes = np.array(geovar_codes)\n",
    "    geovar_obj.n_variants = geovar_obj.geovar_codes.size\n",
    "    geovar_obj.n_populations = geovar_obj.pops.size\n",
    "\n",
    "geovar = GeoVar(bins=[(0,0), (0,0.01), (0.01,0.05), (0.05,0.1), (0.1,1.0)])\n",
    "geovar_codes_streaming_fixed(geovar_obj=geovar, freq_mat_file=\"/quobyte/bmhenngrp/from-lssc0/projects/CAAPA2_functional_annotation/geovar/results/freq_mat_file.csv.gz\") \n",
    "counts_output = geovar.count_geovar_codes()\n",
    "geovar_code_counts = pd.DataFrame({\"codes\":counts_output[0],\"counts\":counts_output[1]})\n",
    "geovar_code_counts.to_csv(\"/quobyte/bmhenngrp/from-lssc0/projects/CAAPA2_functional_annotation/geovar/data/caapa_chr21.txt\", sep=' ', header=False, index=False)\n",
    "\n",
    " \n",
    " "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:python_geovar]",
   "language": "python",
   "name": "conda-env-python_geovar-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
